bionic/flow.py

```diff
@@ -114,22 +113,27 @@ def add_case(self, name, case_key, value):
     def clear_providers(self, names):
         state = self
 
-        # Remember the original protocol for each provider.
-        protocols_by_entity_name = {}
+        # Remember the original provider attributes per entity.
+        attrs_by_entity_name = {}
         for name in names:
             if not state.has_provider(name):
                 continue
             provider = self.get_provider(name)
-            for res_name, res_protocol in zip(
-                    provider.attrs.names, provider.attrs.protocols):
-                protocols_by_entity_name[res_name] = res_protocol
+            attrs_by_entity_name[name] = provider.attrs
 
         # Delete the providers (or fail if not possible).
         state = state.delete_providers(names)
 
+        seen_attrs = set()
+
         # Recreate an empty version of each provider.
-        for res_name, res_protocol in protocols_by_entity_name.items():
-            state = state.create_provider(res_name, res_protocol)
+        for name, attrs in attrs_by_entity_name.items():
+            for res_name, res_protocol in zip(attrs.names, attrs.protocols):
```

Comment:
This seems like it will work, but it's not clear to me why we need an inner loop at all.  Could we just do this?

 ```
for name, attrs in attrs_by_entity_name.items():
    protocol = attrs.protocols[attrs.names.index(name)]
    state = state.create_provider(name, protocol, attrs.docstring)
```

But if I'm missing something and we do need an inner loop, does `seen_attrs` need to contain anything but the name?  I'm not certain we want to be hashing a Protocol object....

Reply:
oh my, we certainly did overcomplicate this.  Thanks, your suggestion makes sense.